pkg/build/controller/controller.go

```diff
@@ -79,21 +83,53 @@ func (bc *BuildController) CancelBuild(build *buildapi.Build) error {
 // HandleBuild deletes pods for cancelled builds and takes new builds and puts
 // them in the pending state after creating a corresponding pod
 func (bc *BuildController) HandleBuild(build *buildapi.Build) error {
-	glog.V(4).Infof("Handling build %s/%s", build.Namespace, build.Name)
+	glog.V(4).Infof("Handling build %s/%s (%s)", build.Namespace, build.Name, build.Status.Phase)
+
+	// When the current build is complete, get the next build and trigger it by
+	// updating the StartTimestamp. That will trigger an event and force the
+	// RetryController to handle it.
+	// This avoids the default delay the RetryController has to schedule next
+	// builds. However this only works if the current build is "serial" build. In
+	// case the build is parallel, we fallback to RetryController to schedule the
+	// next serial build.
+	if buildutil.IsBuildComplete(build) {
+		nextBuild, err := scheduler.GetNextBuild(bc.OSClient.Builds(build.Namespace), build)
+		if err != nil {
+			glog.Infof("Unable to get the next build for %s/%s: %v", build.Namespace, build.Name, err)
+			return nil
+		}
+		if nextBuild == nil {
+			return nil
+		}
+		now := unversioned.Now()
+		nextBuild.Status.StartTimestamp = &now
+		if err := bc.BuildUpdater.Update(nextBuild.Namespace, nextBuild); err != nil && !errors.IsConflict(err) {
+			return err
+		}
+		bc.Recorder.Eventf(nextBuild, kapi.EventTypeNormal, "scheduled", "Scheduling build %s/%s to run as next",
+			nextBuild.Namespace, nextBuild.Name)
+		return nil
+	}
 
 	// A cancelling event was triggered for the build, delete its pod and update build status.
-	if build.Status.Cancelled && build.Status.Phase != buildapi.BuildPhaseCancelled {
+	if build.Status.Phase != buildapi.BuildPhaseCancelled && build.Status.Cancelled {
+		build.Status.Reason = buildapi.StatusReasonCancelBuildFailed
```

Comment:
why are you setting this here?  it should only be set if the cancel attempt failed.